In [ ]:
import pandas as pd
print("hello")

hello


## Pandas Notes
**pd.read_csv()**

used to read a csv file 


Series- 1D labeled array , can hold any data type like int, float, str or python objects. Each element in this has unique label called an index.
Dataframe- 2D labeled data structure in pandas, like a table in sql, the rows have indices and coloumns have labels.


In [ ]:
import pandas as pd
 
data ={
    'name': ['Alice', 'Bob', 'Charlie'],
    "age":[20,10,30],
    "city":["nagpur","mumbai","delhi"]
    } 
df= pd.DataFrame(data)
print(df)

df.to_csv("output.csv",index=False)


      name  age    city
0    Alice   20  nagpur
1      Bob   10  mumbai
2  Charlie   30   delhi


The dataframe is made and if we wan to save it in a file then write the last line and the type of file u want to make and if u don't want the index then write false along with the code.

Data Exploration -1.  understand the data set(rows and coloumns) 2. identify the problems 3. plan next steps .

to display rows of a dataset head(n) and tail(n) are used.If u skip the n then it prints first 5 rows by default.

In [ ]:
import pandas as pd
df= pd.read_csv(r"C:\Users\Saloni Sharma\Downloads\sales_data.csv")
print("Display first 5 rows:")
df.head(5)
print("Display last 5 rows:")
df.tail(5)

Display first 5 rows:
Display last 5 rows:


,Order_ID,Date,Region,Product,Quantity,Unit_Price_INR,Sales_INR
15,1016,2026-05-02,West,Laptop,4,55000,220000
16,1017,2026-05-09,North,Mobile Phone,8,22000,176000
17,1018,2026-05-17,South,Headphones,20,2500,50000
18,1019,2026-06-05,East,Monitor,6,15000,90000
19,1020,2026-06-18,West,Keyboard,12,1800,21600


Now to understand the data , we should now how many rows and columns are there, what type of data it is,where is missing data. 
For that we use info() method which will give a concise summary of the data like number of rows and colns, columns name , datatype,  non null counts, memory usage of the data frame.

In [ ]:
df= pd.read_csv(r"C:\Users\Saloni Sharma\Downloads\sales_data.csv")
print("Displaying the info of the data set:")
print(df.info())

Displaying the info of the data set:
<class 'pandas.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   Order_ID        20 non-null     int64
 1   Date            20 non-null     str  
 2   Region          20 non-null     str  
 3   Product         20 non-null     str  
 4   Quantity        20 non-null     int64
 5   Unit_Price_INR  20 non-null     int64
 6   Sales_INR       20 non-null     int64
dtypes: int64(4), str(3)
memory usage: 1.7 KB
None


In [ ]:
import pandas as pd
 
data ={
    'name': ['Alice', 'Bob', 'Charlie'],
    "age":[20,10,30],
    "city":["nagpur","mumbai","delhi"]
    } 
df=pd.DataFrame(data)
print("display the info of the dataset:")
print(df.info())

display the info of the dataset:
<class 'pandas.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   name    3 non-null      str  
 1   age     3 non-null      int64
 2   city    3 non-null      str  
dtypes: int64(1), str(2)
memory usage: 236.0 bytes
None


Describe Method- It provides a summary of the descriptive statistics for numerical columns in your dataframe.

In [ ]:
# step1 - sample data frame
import pandas as pd

data= { "Name":["raj","sahil","rishi","rohit", "ramesh","sachin"],
        "Age":[20,21,22,23,24,25],
        "salary":[49000,50000,56000,67000,47000,76000],
        "performance score":[90,80,45,26,37,27]
}
df=pd.DataFrame(data)
print("sample DataFrame:")
print(df)
print("descriptive statistics:")
print(df.describe())


sample DataFrame:
     Name  Age  salary  performance score
0     raj   20   49000                 90
1   sahil   21   50000                 80
2   rishi   22   56000                 45
3   rohit   23   67000                 26
4  ramesh   24   47000                 37
5  sachin   25   76000                 27
descriptive statistics:
             Age        salary  performance score
count   6.000000      6.000000           6.000000
mean   22.500000  57500.000000          50.833333
std     1.870829  11606.032914          27.549350
min    20.000000  47000.000000          26.000000
25%    21.250000  49250.000000          29.500000
50%    22.500000  53000.000000          41.000000
75%    23.750000  64250.000000          71.250000
max    25.000000  76000.000000          90.000000


The two outputs are entirely differnet. The describe method have described the data . Count-The no. of non nulls in each column.

Mean- the arithematic average of the value in each column.

Std- (standard deviation)-tells us how much of the values in a column are diff from the average(mean).

For understanding it has two types 

small std - those values which are very close to the average like [49,48,51,52] all are very close to the 50.

large std-like salaries of the employees are very differ and spread out .
Used to measure the consistency like if small std then more consistency.
Used to detect variations like if large std then means large variation in data.

Min - minimum value , smallest value in that column data.

25%- means first quarter of your data,if we sort our numbers from smallest to largest then 25% of data will be below this value . if we sort our data and divide it into 4 groups then 25% will show the min value it can be used to tell you what value separate that lowest 25% data from the rest.eg it can show the salary threshold for the lowest paid employees.eg- to calculate low paying job.

50% also called medain- the middle va;ue of your data when it was sorted. like 22+23/2=22.5 eg - to calculate medium salary paying jobs.

75% tells you about the highest 255 of your data from the rest. eg- to calculate highest paying job

Max- max value in the column.

Shape- used to know about the no. of rows and columns in your dataset. It will return a tuple with two values no. of rows and colns.
column- will give the names of the columns.we can know about structure of the dataset

In [1]:
import pandas as pd

data= { "Name":["raj","sahil","rishi","rohit", "ramesh","sachin"],
        "Age":[20,21,22,23,24,25],
        "salary":[49000,50000,56000,67000,47000,76000],
        "performance score":[90,80,45,26,37,27]
}
df= pd.DataFrame(data)
print(df)
print('Shape:',df.shape)
print('Columns:',df.columns)

     Name  Age  salary  performance score
0     raj   20   49000                 90
1   sahil   21   50000                 80
2   rishi   22   56000                 45
3   rohit   23   67000                 26
4  ramesh   24   47000                 37
5  sachin   25   76000                 27
Shape: (6, 4)
Columns: Index(['Name', 'Age', 'salary', 'performance score'], dtype='object')


 To Select specific column - returns a series and a dataframe.
 column= df["column name"]
 subset=df["column1","column2","..."]

 To Filter rows based on the conditions- boolean indexing is used here. based on a single condition-
 filtered_rows= df[df["Salary"]> 50000]
 
 To Combine multiple conditions-
filtered_rows=df[(df["column1"]> value1) &(df["column2"]<value2)]
 

In [ ]:
import pandas as pd

data= { "Name":["raj","sahil","rishi","rohit", "ramesh","sachin"],
        "Age":[20,21,22,23,24,25],
        "salary":[49000,50000,56000,67000,47000,76000],
        "performance score":[90,80,45,26,37,27]
}
df=pd.DataFrame(data)
print("sample dataframe:")
print(df)
print("Name (single column return series)")
print(df["Name"])
#selecting multiple columns
subset= df[["Name", "salary"]]
print(subset)

sample dataframe:
     Name  Age  salary  performance score
0     raj   20   49000                 90
1   sahil   21   50000                 80
2   rishi   22   56000                 45
3   rohit   23   67000                 26
4  ramesh   24   47000                 37
5  sachin   25   76000                 27
Name (single column return series)
0       raj
1     sahil
2     rishi
3     rohit
4    ramesh
5    sachin
Name: Name, dtype: object
     Name  salary
0     raj   49000
1   sahil   50000
2   rishi   56000
3   rohit   67000
4  ramesh   47000
5  sachin   76000


In [12]:
#to filter rows
import pandas as pd

data= { "Name":["raj","sahil","rishi","rohit", "ramesh","sachin"],
        "Age":[20,21,22,23,24,25],
        "salary":[49000,50000,56000,67000,47000,76000],
        "performance score":[90,80,45,26,37,27]
}
df=pd.DataFrame(data)
#to filter rows ona single condition

high_salary=df[df["salary"]>50000]
print(high_salary)

#with multiple conditions using and which returns if all the conditions are true
high_salary=df[(df["salary"]> 50000)& (df["Age"]<50)]
print('Employee salary and age')
print(high_salary)

# using or condition that will return even if one condition is true 
filtered_or= df[(df["Age"]>30) | (df["performance score"] >40)]
print(filtered_or)

     Name  Age  salary  performance score
2   rishi   22   56000                 45
3   rohit   23   67000                 26
5  sachin   25   76000                 27
Employee salary and age
     Name  Age  salary  performance score
2   rishi   22   56000                 45
3   rohit   23   67000                 26
5  sachin   25   76000                 27
    Name  Age  salary  performance score
0    raj   20   49000                 90
1  sahil   21   50000                 80
2  rishi   22   56000                 45


Modifying data - using two methods 

In [1]:
#Adding columns
import pandas as pd

data= { "Name":["raj","sahil","rishi","rohit", "ramesh","sachin"],
        "Age":[20,21,22,23,24,25],
        "salary":[49000,50000,56000,67000,47000,76000],
        "performance score":[90,80,45,26,37,27]
}
df= pd.DataFrame(data)
#adding columns via assignment using square brackets df["column_name"]= some_data
print(df)

df["bonus"]=df["salary"]*0.1
print(df)

#using insert() method you can add a new column at specific  position or index in the dataframe . we can add the column at a precise location.
#df.insert(loc,"column_name",some_data)
df.insert(0,"Employee_ID",[10,20,30,40,50,60])
print(df)


     Name  Age  salary  performance score
0     raj   20   49000                 90
1   sahil   21   50000                 80
2   rishi   22   56000                 45
3   rohit   23   67000                 26
4  ramesh   24   47000                 37
5  sachin   25   76000                 27
     Name  Age  salary  performance score   bonus
0     raj   20   49000                 90  4900.0
1   sahil   21   50000                 80  5000.0
2   rishi   22   56000                 45  5600.0
3   rohit   23   67000                 26  6700.0
4  ramesh   24   47000                 37  4700.0
5  sachin   25   76000                 27  7600.0
   Employee_ID    Name  Age  salary  performance score   bonus
0           10     raj   20   49000                 90  4900.0
1           20   sahil   21   50000                 80  5000.0
2           30   rishi   22   56000                 45  5600.0
3           40   rohit   23   67000                 26  6700.0
4           50  ramesh   24   47000      

Updating Values using  this .loc[] you can access specific cell , row ,column and can do modification.
syntax- df.loc[row_index,"Column_name"]= new_value
when to use .loc - when u want to update a row, cell or column precisely.

In [3]:
#updating Ram's salary

data= { "Name":["raj","sahil","rishi","rohit", "ramesh","sachin"],
        "Age":[20,21,22,23,24,25],
        "salary":[49000,50000,56000,67000,47000,76000],
        "performance score":[90,80,45,26,37,27]
}
df= pd.DataFrame(data)
df.loc[0,"salary"]=56000
print(df)

     Name  Age  salary  performance score
0     raj   20   56000                 90
1   sahil   21   50000                 80
2   rishi   22   56000                 45
3   rohit   23   67000                 26
4  ramesh   24   47000                 37
5  sachin   25   76000                 27


In [4]:
#if we want to update more than one value of a column 
data= { "Name":["raj","sahil","rishi","rohit", "ramesh","sachin"],
        "Age":[20,21,22,23,24,25],
        "salary":[49000,50000,56000,67000,47000,76000],
        "performance score":[90,80,45,26,37,27]
}
df= pd.DataFrame(data)
#increasing salary by 5%
df["salary"]=df['salary']*1.05
print(df)

     Name  Age   salary  performance score
0     raj   20  51450.0                 90
1   sahil   21  52500.0                 80
2   rishi   22  58800.0                 45
3   rohit   23  70350.0                 26
4  ramesh   24  49350.0                 37
5  sachin   25  79800.0                 27


Removing columns- we use drop method in pandas for it, by this we can delete any row and column from any dataframe.
syntax- df.drop[columns = ["column_name"],inplace= true]
inplace means it will modify the original dataframe directly. 
if  inplace= false, it will return a new dataframe.

In [9]:
data= { "Name":["raj","sahil","rishi","rohit", "ramesh","sachin"],
        "Age":[20,21,22,23,24,25],
        "salary":[49000,50000,56000,67000,47000,76000],
        "performance score":[90,80,45,26,37,27]
}
df= pd.DataFrame(data)
df.drop(columns=["performance score"], inplace=True)
print(df)
df.drop(columns=["Age","salary"],inplace=True)
print(df)

     Name  Age  salary
0     raj   20   49000
1   sahil   21   50000
2   rishi   22   56000
3   rohit   23   67000
4  ramesh   24   47000
5  sachin   25   76000
     Name
0     raj
1   sahil
2   rishi
3   rohit
4  ramesh
5  sachin


Handling missing values-missingndata is represented through "Nan" - not a number , None (for object data types).
Before handling missing data find where it is missing . use isnull() , it will give a boolean value True - NaN is missing 
False - value is present
to find no. of missing values add . sum in the end of isnull
Now , how to handle the missing value -either remove it or fill it .
dropna()- df.dropna(axis=0 , inplace =True)
axis=0 , means dropping the row with missing value , axis=1 dropping column with missing value.
inplace = true , means it will give the original dataframe after modifying it .

fillna()-fillna(value, inplace=True)

Interpolation -interpolate() a technique through which you can fill estimated values in missing places, it estimates missing values in numerical columns.
why use this- preserves data integrity
-smooth trends (where continuity matters)
-avoid data loss
Methods- linear ,polynomial, time ,quadratic, cubic, index , values, pad , nearest,akima.

Where to use - time series data , numeric data with trends, avoid dropping rows .
 data remains structured , but can't work with categorical data


In [3]:
import pandas as pd
data= { "Name":["raj",None,"rishi","rohit", "ramesh","sachin"],
        "Age":[20,None,22,23,24,25],
        "salary":[49000,None,56000,67000,47000,76000],
        "performance score":[90,None,45,26,37,27]
}
df= pd.DataFrame(data)
print(df)
print(df.isnull())
print(df.isnull().sum())

     Name   Age   salary  performance score
0     raj  20.0  49000.0               90.0
1    None   NaN      NaN                NaN
2   rishi  22.0  56000.0               45.0
3   rohit  23.0  67000.0               26.0
4  ramesh  24.0  47000.0               37.0
5  sachin  25.0  76000.0               27.0
    Name    Age  salary  performance score
0  False  False   False              False
1   True   True    True               True
2  False  False   False              False
3  False  False   False              False
4  False  False   False              False
5  False  False   False              False
Name                 1
Age                  1
salary               1
performance score    1
dtype: int64


In [ ]:
import pandas as pd
data= { "Name":["raj",None,"rishi","rohit", "ramesh","sachin"],
        "Age":[20,None,22,23,24,25],
        "salary":[49000,None,56000,67000,47000,76000],
        "performance score":[90,None,45,26,37,27]
}
df= pd.DataFrame(data)
print(df)
df.dropna(inplace=True)# it will remove the nan 
print(df)

     Name   Age   salary  performance score
0     raj  20.0  49000.0               90.0
1    None   NaN      NaN                NaN
2   rishi  22.0  56000.0               45.0
3   rohit  23.0  67000.0               26.0
4  ramesh  24.0  47000.0               37.0
5  sachin  25.0  76000.0               27.0
     Name   Age   salary  performance score
0     raj  20.0  49000.0               90.0
2   rishi  22.0  56000.0               45.0
3   rohit  23.0  67000.0               26.0
4  ramesh  24.0  47000.0               37.0
5  sachin  25.0  76000.0               27.0


In [5]:
import pandas as pd
data= { "Name":["raj",None,"rishi","rohit", "ramesh","sachin"],
        "Age":[20,None,22,23,24,25],
        "salary":[49000,None,56000,67000,47000,76000],
        "performance score":[90,None,45,26,37,27]
}
df= pd.DataFrame(data)
print(df)
df.fillna(0, inplace=True)
print(df)

     Name   Age   salary  performance score
0     raj  20.0  49000.0               90.0
1    None   NaN      NaN                NaN
2   rishi  22.0  56000.0               45.0
3   rohit  23.0  67000.0               26.0
4  ramesh  24.0  47000.0               37.0
5  sachin  25.0  76000.0               27.0
     Name   Age   salary  performance score
0     raj  20.0  49000.0               90.0
1       0   0.0      0.0                0.0
2   rishi  22.0  56000.0               45.0
3   rohit  23.0  67000.0               26.0
4  ramesh  24.0  47000.0               37.0
5  sachin  25.0  76000.0               27.0


In [6]:
# to fill calculated value in the Nan cells
import pandas as pd
data= { "Name":["raj",None,"rishi","rohit", "ramesh","sachin"],
        "Age":[20,None,22,23,24,25],
        "salary":[49000,None,56000,67000,47000,76000],
        "performance score":[90,None,45,26,37,27]
}
df= pd.DataFrame(data)
print(df)
df["Age"].fillna(df["Age"].mean() , inplace =True)
df["salary"].fillna(df["salary"].mean(), inplace =True)
df["performance score"].fillna(df["performance score"].mean(),inplace =True)
print(df)

     Name   Age   salary  performance score
0     raj  20.0  49000.0               90.0
1    None   NaN      NaN                NaN
2   rishi  22.0  56000.0               45.0
3   rohit  23.0  67000.0               26.0
4  ramesh  24.0  47000.0               37.0
5  sachin  25.0  76000.0               27.0
     Name   Age   salary  performance score
0     raj  20.0  49000.0               90.0
1    None  22.8  59000.0               45.0
2   rishi  22.0  56000.0               45.0
3   rohit  23.0  67000.0               26.0
4  ramesh  24.0  47000.0               37.0
5  sachin  25.0  76000.0               27.0


C:\Users\Saloni Sharma\AppData\Local\Temp\ipykernel_23200\3270595350.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Age"].fillna(df["Age"].mean() , inplace =True)
C:\Users\Saloni Sharma\AppData\Local\Temp\ipykernel_23200\3270595350.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves

In [7]:
import pandas as pd
data= { "Name":["raj","shyam","rishi","rohit", "ramesh","sachin"],
        "Age":[20,None,22,23,24,25],
        "salary":[49000,None,56000,67000,47000,76000],
        "performance score":[90,None,45,26,37,27]
}
df= pd.DataFrame(data)
print(df)

df.interpolate(method='linear',axis=0 , inplace=True)
print(df)

     Name   Age   salary  performance score
0     raj  20.0  49000.0               90.0
1   shyam   NaN      NaN                NaN
2   rishi  22.0  56000.0               45.0
3   rohit  23.0  67000.0               26.0
4  ramesh  24.0  47000.0               37.0
5  sachin  25.0  76000.0               27.0
     Name   Age   salary  performance score
0     raj  20.0  49000.0               90.0
1   shyam  21.0  52500.0               67.5
2   rishi  22.0  56000.0               45.0
3   rohit  23.0  67000.0               26.0
4  ramesh  24.0  47000.0               37.0
5  sachin  25.0  76000.0               27.0


C:\Users\Saloni Sharma\AppData\Local\Temp\ipykernel_23200\2180595686.py:10: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  df.interpolate(method='linear',axis=0 , inplace=True)


Sorting and aggreagtion- hepls arranging data in a meaningul manner.

Sorting-arranging rows in a dataframe based on the one or more columns. 
sorting data in one column-sort_values() method is used.
syntax-
df.sort_values(by="Column Name",True/False,inplace=True)

Aggregating-involves calculating summary statistics .
summary statistics- provides numeric summaries of columns min,max,average.
syntax- df['column name"].min

In [11]:
import pandas as pd
data={
    "Name":["raj","shyam","rishi"],
    "age":[20,24,25],
    "salary":[49000,56000,67000],
}
df=pd.DataFrame(data)
df.sort_values(by="age",ascending=False ,inplace=True)
print("sorting age by descending order:")
print(df)
 #fpr multiple columns we can pass a list of column names to the by parameter of the sort_values() method.
df.sort_values(by=["age","salary"],ascending=[True, False],inplace=True)
print(df)

sorting age by descending order:
    Name  age  salary
2  rishi   25   67000
1  shyam   24   56000
0    raj   20   49000
    Name  age  salary
0    raj   20   49000
1  shyam   24   56000
2  rishi   25   67000


In [12]:
import pandas as pd
data={
    "Name":["raj","shyam","rishi"],
    "age":[20,24,25],
    "salary":[49000,56000,67000],
}
df=pd.DataFrame(data)
avg_salary=df["salary"].mean()
print("average_salary:",avg_salary)


average_salary: 57333.333333333336


Grouping- dividing dataset  into grps . in the code first the data is grouped and then aggregation is apllied on it.

Grouping in multiple columns- add a list of column u want to aggreagte .
Common aggregation functions- sum(), mean(), max(),min(), std().

In [15]:
import pandas as pd
data={
    "Name":["raj","shyam","rishi","kishan","rahul"],
    "age":[20,24,25,24,20],
    "salary":[49000,56000,67000,78000,89000],
    
}
df= pd.DataFrame(data)
grouped= df.groupby("age")["salary"].sum()
print(grouped)# it will give the sum of salary for each age group

grouped=df.groupby(["age","Name"])["salary"].sum()
print(grouped)

age
20    138000
24    134000
25     67000
Name: salary, dtype: int64
age  Name  
20   rahul     89000
     raj       49000
24   kishan    78000
     shyam     56000
25   rishi     67000
Name: salary, dtype: int64


Merging and joining - merging - joining  rows of two or more than two dataframes.
syntax - pd.merge(df1,df2, on="column name", how="type of join")

In [22]:
import pandas as pd 
#customer dataframe 
df_customers = pd.DataFrame(
    {'customer_id': [1, 2, 3],
     'Name':['Alice', 'Bob', 'Charlie'],
     }
)

#order dataframe
df_orders=pd.DataFrame({
    'customer_id': [1, 2, 4],
    'orderamount': [200, 240, 367],
    
})
#merge 
df_merged=pd.merge(df_customers,df_orders,on='customer_id',how="inner")
print("inner join")
print(df_merged)

#outer will merge all the rows and will fill missing values with nan
df_merged=pd.merge(df_customers,df_orders,on='customer_id',how="outer")
print("outer join")
print(df_merged)

#left join will keep all the rows from the left dataframe and fill missing values with nan for the right dataframe
df_merged=pd.merge(df_customers,df_orders,on='customer_id',how="left")
print("left join")
print(df_merged)

#right join will keep all the rows from the right dataframe and fill missing values with nan for the left dataframe
df_merged=pd.merge(df_customers,df_orders,on='customer_id',how="right")
print("right join")
print(df_merged)
#cross join will return the cartesian product of both dataframes


inner join
   customer_id   Name  orderamount
0            1  Alice          200
1            2    Bob          240
outer join
   customer_id     Name  orderamount
0            1    Alice        200.0
1            2      Bob        240.0
2            3  Charlie          NaN
3            4      NaN        367.0
left join
   customer_id     Name  orderamount
0            1    Alice        200.0
1            2      Bob        240.0
2            3  Charlie          NaN
right join
   customer_id   Name  orderamount
0            1  Alice          200
1            2    Bob          240
2            4    NaN          367


Concatenation- combinig dataframes vertically or horizontally.
pd.concate([df1,df2],axis=0 , ignore_index=True)
axis =0 , rows wise 
axis=1 , column wise combine
ignore_index = true , reset index 

In [25]:
#vertically
import pandas as pd

df_region1= pd.DataFrame({
    'customerID':[1,2],
    'name':['gopal','raju']
})
    

df_region2= pd.DataFrame({
    'customerID':[3,4],
    'name':['shiva','raaj']
})

df_concat = pd.concat([df_region1,df_region2] ,ignore_index=True)
print(df_concat)

df_concat = pd.concat([df_region1,df_region2] ,axis =1 ,ignore_index=True)
print(df_concat)


   customerID   name
0           1  gopal
1           2   raju
2           3  shiva
3           4   raaj
   0      1  2      3
0  1  gopal  3  shiva
1  2   raju  4   raaj
